presto-main/src/main/java/com/facebook/presto/operator/spiller/BinaryFileSpiller.java

```diff
@@ -0,0 +1,152 @@
+
+    static {
+        SPILL_PATH.toFile().mkdirs();
+    }
+
+    private final Path targetDirectory;
+    private final Closer closer = Closer.create();
+    private final BlockEncodingSerde blockEncodingSerde;
+
+    private int spillsCount = 0;
+    private final ListeningExecutorService executor;
+
+    public BinaryFileSpiller(BlockEncodingSerde blockEncodingSerde, ListeningExecutorService executor)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        try {
+            this.targetDirectory = Files.createTempDirectory(SPILL_PATH, "presto-spill");
+        }
+        catch (IOException e) {
+            throw new PrestoException(INTERNAL_ERROR, "Failed to create spill directory", e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> spill(Iterator<Page> pageIterator)
+    {
+        Path spillPath = getPath(spillsCount++);
+
+        return MoreFutures.toCompletableFuture(executor.submit(
+                () -> {
+                    try {
+                        writePages(pageIterator, spillPath);
+                    }
+                    catch (IOException e) {
+                        throw new PrestoException(INTERNAL_ERROR, "Failed to spill pages", e);
+                    }
+                }
+        ));
+    }
+
+    private void writePages(Iterator<Page> pageIterator, Path spillPath)
+            throws IOException
+    {
+        try (OutputStream output = new BufferedOutputStream(new FileOutputStream(spillPath.toFile()))) {
+            PagesSerde.writePages(blockEncodingSerde, new OutputStreamSliceOutput(output), pageIterator);
+        }
+        catch (RuntimeIOException | IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @Override
+    public List<Iterator<Page>> getSpills()
+    {
+        try {
+            ImmutableList.Builder<Iterator<Page>> spills = ImmutableList.builder();
+            for (int spillNumber = 0; spillNumber < spillsCount; spillNumber++) {
+                Path spillPath = getPath(spillNumber);
+                spills.add(readPages(spillPath));
+            }
+
+            return spills.build();
+        }
+        catch (IOException e) {
+            throw new PrestoException(INTERNAL_ERROR, "Failed to read spilled pages", e);
+        }
+    }
+
+    private Iterator<Page> readPages(Path spillPath)
+            throws IOException
+    {
+        InputStream input = new BufferedInputStream(new FileInputStream(spillPath.toFile()));
+        closer.register(input);
+        return PagesSerde.readPages(blockEncodingSerde, new InputStreamSliceInput(input));
+    }
+
+    @Override
+    public void close()
+    {
+        try (Stream<Path> list = Files.list(targetDirectory)) {
+            closer.close();
+            for (Path path : list.collect(toList())) {
+                Files.delete(path);
```

Comment:
Then we would have to deal with handling of multiple exceptions. Do we hide all but first exception? At least for now I would stick with this as it is.